File: labs/02-consumer-groups/consumer.py
import os
import time

from lab_kafka.client import consumer, decode_message, topic_name


def main() -> None:
    topic = topic_name("orders")
    group_id = os.getenv("GROUP_ID", "order-workers")
    worker_name = os.getenv("WORKER_NAME", f"worker-{os.getpid()}")
    kafka_consumer = consumer(group_id)
    kafka_consumer.subscribe([topic])

    print(f'{worker_name} consuming {topic} as group "{group_id}".')

    try:
        while True:
            message = kafka_consumer.poll(1.0)

            if message is None:
                continue

            if message.error():
                print(message.error())
                continue

            event = decode_message(message)
            time.sleep(0.5)

            print(
                {
                    "workerName": worker_name,
                    "partition": message.partition(),
                    "offset": message.offset(),
                    "orderId": event["value"]["orderId"],
                    "customerId": event["value"]["customerId"],
                }
            )
            kafka_consumer.commit(message=message, asynchronous=False)
    except KeyboardInterrupt:
        print("Stopping consumer.")
    finally:
        kafka_consumer.close()


if __name__ == "__main__":
    main()

File: labs/02-consumer-groups/consumer.js
import { decodeMessage, kafka, topicName } from "../../lib/kafka.js";

const topic = topicName("orders");
const groupId = process.env.GROUP_ID ?? "order-workers";
const workerName = process.env.WORKER_NAME ?? `worker-${process.pid}`;
const consumer = kafka.consumer({ groupId });

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

process.on("SIGINT", async () => {
  await consumer.disconnect();
  process.exit(0);
});

await consumer.connect();
await consumer.subscribe({ topic, fromBeginning: true });

console.log(`${workerName} consuming ${topic} as group "${groupId}".`);

await consumer.run({
  eachMessage: async ({ partition, message }) => {
    const event = decodeMessage(message);
    await sleep(500);

    console.log({
      workerName,
      partition,
      offset: message.offset,
      orderId: event.value.orderId,
      customerId: event.value.customerId
    });
  }
});

